### DISCUSSION TITLE
Flow & Loop

### DISCUSSION THREAD

**Discussion by imknight** [directus/directus#15720] [Archive]
How do you create a flow that can loop action for all the return records?
for example, I pull all users who have the status "progress" how do I loop through these users to send them emails each?

**Comment by br41nslug** [comment]
If you pass the array of results to a new Flow using the "trigger flow" operation it should loop over the array and trigger the second flow once for each item

**Reply by leon-marzahn** [reply]
Just a quick question, how would you do that with using $last for example?

**Reply by david-zacharias** [reply]
I have also read about this nice behaviour in the docs. But I also struggle on how to pass a non-hard coded array as payload to another flow? @br41nslug can you provide or point uns to an example?

**Reply by br41nslug** [reply]
Giving an example is tricky without knowing something about the datastructure you're working with. However the trick generally is to **only** pass the array to the "trigger flow" operator.
Lets say you have a context object of:
```json
{
  "$last": {
    "thisIsAnArray": [
      { "item": 1 },
      { "item": 2 },
      { "item": 3 },
    ]
  }
}
```
Then passing `{{ $last.thisIsAnArray }}`
@leon-marzahn @david-zacharias

**Reply by imknight** [reply]
yup working, just need to ensure the payload that trigger is array will do

**Reply by eidermaen** [reply]
For me it is somehow not working. My structure is like:
```json
{
  "$last": {
	"data": [
		{
		  "title": "someTitle",
		  "date": "someDate",
		  "description": "Some description"
		},
		{
		  "title": "someOtherTitle",
		  "date": "someOtherDate",
		  "description": "Some other description"
		}
	]
  }
}
```

Inside the Trigger Flow operation I then pass the following as payload:
```json
[
    "{{ $last.data }}"
]
```
I also tried 
```
{{ $last.data }}
```
but neither of them works. The called flow is only executed once but not `$last.data.length` times. What am I doing wrong?
Or is it related to how I handle the payload in the called flow?
Thank you in advance!

**Reply by david-zacharias** [reply]
Sorry for my delayed feedback and thanks @br41nslug for your quick reply.

>Then passing {{ $last.thisIsAnArray }}

The exact syntax for the `Payload`-Field was our main problem, which seems to be same for @eidermaen 

We got it working like so:

<img width="570" alt="Bildschirmfoto 2022-11-24 um 16 17 33" src="https://user-images.githubusercontent.com/1903179/203818534-551c284a-82bc-4653-8c1b-0de6402d5fff.png">

In our case, the `$last` contains data we Read first:

<img width="1045" alt="Bildschirmfoto 2022-11-24 um 16 20 49" src="https://user-images.githubusercontent.com/1903179/203818729-695bf489-5c60-4f34-988a-ec02a26adc9a.png">

So, I guess for you @eidermaen the syntax is `"{{$last.data}}"` supposed, `data` is an array

**Reply by eidermaen** [reply]
Thank's a lot @david-zacharias! That did the trick! So I had to edit the payload as raw value and entered `"{{ $last.data }}"`. This way the triggered flow was executed `$last.data.length` times ;)

**Comment by Mehdi-YC** [comment]
Hi, i have almost the same problem, i have a string value and i need to transform it to an int,i gotthe values with read  data, and i want to pass it to updatedata one by one, is it possible within the same flow ?

**Comment by sigzegv** [comment]
Is it supposed to work on a specific version ? I'm on an older v10.8 and I tried the following : 
In the first flow: 
=> I made Read Data action that returns an array from a collection
=> then a create a Trigger flow action with "{{ "data": $last }}" as payload

The second flow is Another flow:
=> Then I made a Read data with nothing but it gets the whole previous array at once. How can I read the array data element by element ?

**Reply by br41nslug** [reply]
Make sure the root object you're passing on is an array `[ item_1, item_2 ]` not an array nested in an object like `{ data: [...] }`

**Reply by sigzegv** [reply]
If I just return the payload as "{{ $last }}" I get a json error like "unexpected token i in JSON position 4" . But knowing that my $last is directly from a Read Data, format should be correct..

**Reply by br41nslug** [reply]
That error sounds like a quote escaping issue when trying to parse it as json

**Reply by sigzegv** [reply]
An escaping issue from the collection's data maybe ? (resultset come from a collection without modifications)

**Comment by sigzegv** [comment]
How do you catch the data from the another flow action starting the new flow? seems the read data action is not suitable for that as it only reads from a database collection.

**Reply by david-zacharias** [reply]
The data of actions (= operations) is can be accessed in following operations by the operations name: [The Data Chain in the docs](https://docs.directus.io/app/flows.html#the-data-chain). Say you have a read operation with name `books` you can access the data via `{{ books }}` in following operations.
